### PR TITLE
Jacob/rs 8 delete account request feedback

### PR DESCRIPTION
**Pull request by jacobperia**
## 📜 Description
Updated Delete section in user settings
New model to store user feedback on deleting account.
Updated delete action to create feedback record and handle wrong password error.

## 🖼 Proof of Work
SCR-20230616-ooy

image

SCR-20230616-op2

### CODE REVIEW

File: app/views/devise/registrations/edit.html.haml [hunk 1]
```diff
@@ -66,6 +66,22 @@
       %p.mt-1.p Danger zone! Once you delete you account, there is no going back.
 
     .card.md:col-span-2
-      .card-body
+      .card-body{"x-data" => "{ button: true, open: false }"}
```
Comment (danielpaul) [inline app/views/devise/registrations/edit.html.haml:69]: why do we need button and open? is it not just one action and then it opens? 

how is this handed where the password the user entered is wrong?

File: app/views/devise/registrations/edit.html.haml [hunk 2]
```diff
@@ -66,6 +66,22 @@
...
         .space-y-6
-          = button_to "Delete Account", registration_path(resource_name), {data: { turbo_confirm: "Are you sure?" }, class: 'text-danger text-sm mr-6', method: :delete}
+          %div{ "x-show" => "button" }
+            .text-sm.text-neutral-500.mb-2 Delete your account. This action cannot be undone and all your data will be deleted from our system.
+            %button{ class: 'text-danger text-sm mr-6', method: :delete, "x-on:click" => "open = ! open; button = ! button" } Delete Account
+        = form_for(resource, as: resource_name, url: registration_path(resource_name), builder: RapidRailsFormBuilder, html: { method: :delete }) do |f|
+          .text-black{"x-show" => "open"}
```
Comment (danielpaul) [inline app/views/devise/registrations/edit.html.haml:75]: don't do text-black or any colors like this. we have base styles. you should just use `%p.p` (p also being the class). This handles dark mode, etc. 

The current code does not work for dark mode. 

image

File: app/views/devise/registrations/edit.html.haml [hunk 3]
```diff
@@ -66,6 +66,22 @@
...
+            %p You are about to delete your account! This action can not be undone and all your data will be deleted from our system.
+            %br
+            %p.mb-6
+              We are sorry to see you go #{@user.full_name}, if there was one thing that you could change about our service, what would that be? 
+              This feedback is critical to our continued improvement. Thanks so much!
+
+              .space-y-6
+                = f.password_field :current_password, placeholder: "Your current password", autocomplete: "current-password", required: true
+                = f.text_field :feedback, placeholder: "Please tell us why you are cancelling. Any kind of feedback is appreciated and helpful to us.", autocomplete: 'no'
```
Comment (danielpaul) [inline app/views/devise/registrations/edit.html.haml:84]: make this a text area

File: app/controllers/registrations_controller.rb [hunk 4]
```diff
@@ -17,10 +17,16 @@ def cancel_email_change!
...
-    set_flash_message! :notice, :destroyed
-    respond_with_navigational(resource) { redirect_to after_sign_out_path_for(resource_name), status: Devise.responder.redirect_status }
+    if resource.valid_password?(user_delete_params[:current_password])
+      resource.user_account_feedbacks.create(user: resource, feedback: user_delete_params[:feedback]) if user_delete_params[:feedback]
+      resource.discard
+      Devise.sign_out_all_scopes ? sign_out : sign_out(resource_name)
+      set_flash_message! :notice, :destroyed
+      respond_with_navigational(resource) { redirect_to after_sign_out_path_for(resource_name), status: Devise.responder.redirect_status }
+    else
+      flash.now[:alert] = "Invalid password entered. Please try again."
```
Comment (danielpaul) [inline app/controllers/registrations_controller.rb:27]: do a toast flash message. example on how to use this is in the dashboard controller.

File: app/controllers/registrations_controller.rb [hunk 5]
```diff
@@ -17,10 +17,16 @@ def cancel_email_change!
   end
 
   def destroy
-    resource.discard
-    Devise.sign_out_all_scopes ? sign_out : sign_out(resource_name)
-    set_flash_message! :notice, :destroyed
-    respond_with_navigational(resource) { redirect_to after_sign_out_path_for(resource_name), status: Devise.responder.redirect_status }
+    if resource.valid_password?(user_delete_params[:current_password])
+      resource.user_account_feedbacks.create(user: resource, feedback: user_delete_params[:feedback]) if user_delete_params[:feedback]
```
Comment (danielpaul) [inline app/controllers/registrations_controller.rb:21]: do we need an if condition here? we would create the record regardless of if there is feedback or not. the feedback can be nil but the record still needs to be made so we know when we did this to the user

File: app/models/user.rb [hunk 6]
```diff
@@ -41,7 +41,10 @@ class User < ApplicationRecord
   include User::Omniauthable
   include User::Onboarding
 
+  attr_accessor :feedback
```
Comment (danielpaul) [inline app/models/user.rb:44]:
```suggestion
  attr_accessor :delete_account_feedback
```

Comment (danielpaul) [reply app/models/user.rb:44]: why do we need this here? is this if there are errors in the password and we render the same text that the user typed out?

**Review by danielpaul — CHANGES_REQUESTED** [review]

File: app/models/user.rb (hunk 6 above)
Comment (jacobperia) [reply app/models/user.rb:44]: > why do we need this here? is this if there are errors in the password and we render the same text that the user typed out?

i forgot i could use **fields_for** and nested attributes. have replaced it now.

File: app/controllers/registrations_controller.rb (hunk 5 above)
Comment (jacobperia) [reply app/controllers/registrations_controller.rb:21]: > do we need an if condition here? we would create the record regardless of if there is feedback or not. the feedback can be nil but the record still needs to be made so we know when we did this to the user

the condition is only to check if the password is valid

File: app/views/devise/registrations/edit.html.haml (hunk 1 above)
Comment (jacobperia) [reply app/views/devise/registrations/edit.html.haml:69]: > why do we need button and open? is it not just one action and then it opens?
> 
> how is this handed where the password the user entered is wrong?

On clicking the button 'Delete Feedback', we need to hide the same button before we can show the other text. That's why we have two different components in Alpine

**Review by danielpaul — APPROVED** [review]

**Pull request merged**